src/realm.js

```diff
@@ -453,6 +461,99 @@ export class Realm {
     }
   }
 
+  evaluateWithUndoForDiagnostic(f: () => Value): CompilerDiagnostic | Value {
+    if (!this.useAbstractInterpretation) return f();
+    let savedHandler = this.errorHandler;
+    let diagnostic;
+    try {
+      this.errorHandler = d => {
+        diagnostic = d;
+        return "Fail";
+      };
+      let effects = this.evaluateForEffects(f);
+      this.applyEffects(effects);
+      let resultVal = effects[0];
+      if (resultVal instanceof AbruptCompletion) throw resultVal;
+      if (resultVal instanceof PossiblyNormalCompletion) {
+        // in this case one of the branches may complete abruptly, which means that
+        // not all control flow branches join into one flow at this point.
+        // Consequently we have to continue tracking changes until the point where
+        // all the branches come together into one.
+        resultVal = this.composeWithSavedCompletion(resultVal);
+      }
+      invariant(resultVal instanceof Value);
+      return resultVal;
+    } catch (e) {
+      if (diagnostic !== undefined) return diagnostic;
+      throw e;
+    } finally {
+      this.errorHandler = savedHandler;
+    }
+  }
+
+  evaluateForFixpointEffects(
+    loopContinueTest: () => Value,
+    loopBody: () => EvaluationResult
+  ): void | [Effects, Effects] {
+    try {
+      let effects1 = this.evaluateForEffects((loopBody: any));
+      while (true) {
+        this.restoreBindings(effects1[2]);
+        this.restoreProperties(effects1[3]);
+        let effects2 = this.evaluateForEffects(() => {
+          let test = loopContinueTest();
+          if (!(test instanceof AbstractValue)) throw new FatalError("loop terminates before fixed point");
+          return (loopBody(): any);
+        });
+        this.restoreBindings(effects1[2]);
+        this.restoreProperties(effects1[3]);
+        if (Widen.containsEffects(effects1, effects2)) {
+          // effects1 includes every value present in effects2, so doing another iteration using effects2 will not
+          // result in any more values being added to abstract domains and hence a fixpoint has been reached.
+          let [, , bindings1, pbindings1] = effects1;
+          if (pbindings1.size > 0) return undefined;
```

Comment:
Does `undefined` mean that we cannot reach a fix point?

Reply:
It means that we give up on trying to reach a fix point. That is a temporary restriction. I plan to deal with properties eventually, but that is a bit harder and will make the pull request too large for review.